### PR TITLE
refactor: migrate preview admin form endpoint to Typescript

### PR DESCRIPTION
**Pull request by karrui**
## Problem

Migrates `:formId/adminform/preview` endpoint to Typescript. 

The choice of public fields to display is now in each database model's domain; see:
- `UserModel.getPublicView`
- `AgencyModel.getPublicView`
- `FormModel.getPublicView`

## Solution

**Features**:

- feat(AdminFormCtl): add (and use) handlePreviewAdminForm handler fn
- feat(Form): replace util fn with form model method getPublicView

## Tests

Controller tests have been added.
add tests for FormModel#getPublicView instance method
add tests for FormSvc#getFormPublicView

### CODE REVIEW

File: src/app/modules/form/admin-form/admin-form.controller.ts [hunk 1]
```diff
@@ -104,6 +104,56 @@ export const handleGetAdminForm: RequestHandler<{ formId: string }> = (
...
+      .andThen((user) =>
+        // Step 2: Check whether user has read permissions to form
+        AuthService.getFormAfterPermissionChecks({
+          user,
+          formId,
+          level: PermissionLevel.Read,
+        }),
+      )
+      // Step 3: Remove private details from form for previewing.
+      .map(removePrivateDetailsFromForm)
```
Comment (liangyuanruo) [inline src/app/modules/form/admin-form/admin-form.controller.ts:136]: technically not part of this PR, but I just wanted to point out that this is probably better modeled as an instance method on each form object, so that one can call `form.getPublicView()`.

Comment (liangyuanruo) [reply src/app/modules/form/admin-form/admin-form.controller.ts:136]: it'll also get rid of the awkward util file that always contains functions that we're not sure where else to put

Comment (karrui) [reply src/app/modules/form/admin-form/admin-form.controller.ts:136]: I'll update it in this PR

Comment (liangyuanruo) [reply src/app/modules/form/admin-form/admin-form.controller.ts:136]: ok, will approve first

**Review by liangyuanruo — APPROVED** [review]

**Comment by karrui** [comment]
Rerequesting review since the extraction of form instance method was slightly more complex than it seemed

File: src/app/models/form.server.model.ts [hunk 2]
```diff
@@ -61,6 +62,33 @@ import getUserModel from './user.server.model'
...
+
+const POPULATE_FORM_PARAMS = {
+  path: 'admin',
+  // Remove irrelevant keys from populated fields of form admin and agency.
+  select: '-__v -created -lastModified -updatedAt -lastAccessed',
+  populate: {
+    path: 'agency',
+    select: '-__v -created -lastModified -updatedAt',
+  },
+}
```
Comment (liangyuanruo) [inline src/app/models/form.server.model.ts:90]: could we use a whitelist approach instead of a blacklist, so that the code doesn't become out of date if new keys are added to agency or user collections? i.e. explicitly select the keys to include

File: src/app/models/form.server.model.ts [hunk 3]
```diff
@@ -408,6 +436,18 @@ const compileFormModel = (db: Mongoose): IFormModel => {
...
+  FormSchema.methods.getPublicView = async function (this: IFormSchema) {
+    // Populate form if not populated yet.
+    if (!this.populated('admin')) {
+      await this.populate(POPULATE_FORM_PARAMS).execPopulate()
+    }
+
+    return {
+      ...(pick(this, FORM_PUBLIC_FIELDS) as PublicFormValues),
+      admin: pick(this.admin, 'agency'),
+    }
```
Comment (liangyuanruo) [inline src/app/models/form.server.model.ts:448]: seems like which keys to show are defined both in `POPULATE_FORM_PARAMS` and over here using lodash's `pick`. would it be possible to only define them above in one place, then merely call `return this.toObject()` here?

**Review by mantariksh — APPROVED** [review]
lgtm other than yr's comments!

File: src/app/models/form.server.model.ts [hunk 4]
```diff
@@ -463,15 +503,7 @@ const compileFormModel = (db: Mongoose): IFormModel => {
...
-    return this.findById(formId).populate({
-      path: 'admin',
-      // Remove irrelevant keys from populated fields of form admin and agency.
-      select: '-__v -created -lastModified -updatedAt -lastAccessed',
-      populate: {
-        path: 'agency',
-        select: '-__v -created -lastModified -updatedAt',
-      },
-    })
+    return this.findById(formId).populate(POPULATE_FORM_PARAMS)
```
Comment (liangyuanruo) [inline src/app/models/form.server.model.ts:506]: I checked the use of `getFullFormById` and it's only used for internal server-side functions; I think we can simplify the code to= disregard the public view of a form:

```suggestion
    return this.findById(formId).populate({
      path: 'admin',
      populate: {
        path: 'agency',
      },
    }))
```

Comment (karrui) [reply src/app/models/form.server.model.ts:506]: Good suggestion, we can always use `getPublicView` instance method to pick it out if we need to yea?

File: src/app/modules/form/form.service.ts [hunk 5]
```diff
@@ -120,3 +123,25 @@ export const isFormPublic = (
...
+      message: 'Error retrieving form public view',
+      meta: {
+        action: 'getFormPublicView',
+        formId: form._id,
+      },
+      error,
+    })
+    return transformMongoError(error)
+  })
+}
```
Comment (liangyuanruo) [inline src/app/modules/form/form.service.ts:147]: This service method is only needed because `form.getPublicView()` implementation is async, which in turn is because it contains the implementation for populating the `admin` key. Could `form.getPublicView()` be implemented such that it is agnostic to whether the form object is populated or not?

In fact one could even implement `agency.getPublicView()` and `user.getPublicView()`

**Review by liangyuanruo — CHANGES_REQUESTED** [review]
Some comments for your consideration, please.

File: src/app/models/form.server.model.ts (hunk 3 above)
Comment (liangyuanruo) [reply src/app/models/form.server.model.ts:448]: (do take the above under advisement given other conflicting suggestions)

File: src/app/models/form.server.model.ts (hunk 4 above)
Comment (liangyuanruo) [reply src/app/models/form.server.model.ts:506]: yeah, and i think we can lose the `getFormPublicView` service method after that

File: src/app/modules/form/form.service.ts (hunk 5 above)
Comment (karrui) [reply src/app/modules/form/form.service.ts:147]: so getPublicView just returns the subset of the form document without populating (if its not populated)?

so for non-populated form `nonPopulatedForm.getPublicView()` will return an unpopulated agency id, yes?

**Comment by karrui** [comment]
Ready for rereview

**Comment by liangyuanruo** [comment]
Closing this PR for the time being until @karrui 's involvement on other project is reduced.

**Review by mantariksh — APPROVED** [review]
lgtm

File: src/app/modules/form/form.utils.ts [hunk 6]
```diff
@@ -9,56 +6,27 @@ import {
...
+// into an array of emails
+export const transformEmails = (v: string | string[]): string[] => {
+  // Cases
+  // ['test@hotmail.com'] => ['test@hotmail.com'] ~ unchanged
+  // ['test@hotmail.com', 'test@gmail.com'] => ['test@hotmail.com', 'test@gmail.com'] ~ unchanged
+  // ['test@hotmail.com, test@gmail.com'] => ['test@hotmail.com', 'test@gmail.com']
+  // ['test@hotmail.com, test@gmail.com', 'test@yahoo.com'] => ['test@hotmail.com', 'test@gmail.com', 'test@yahoo.com']
+  // 'test@hotmail.com, test@gmail.com' => ['test@hotmail.com', 'test@gmail.com']
+  if (Array.isArray(v)) {
+    return transformEmailString(v.join(','))
```
Comment (liangyuanruo) [inline src/app/modules/form/form.utils.ts:27]: I think this is more elegant and easier to comprehend

```suggestion
    return v.flatMap(transformEmailString)
```

**Review by liangyuanruo — APPROVED** [review]
I notice the consistent use of `getPublicView()` which is great, so let's create a new interface in order to embed that consistency:

```typescript
interface PublicView<T> {
  getPublicView(): T
}

export interface IAgencySchema extends IAgency, Document, PublicView<PublicAgency> {...}

```
